### PR TITLE
Auto switching light/dark theme according to system settings

### LINKED ISSUE #4970: Respect `prefers-color-scheme` of the underlying OS
### Check for existing issues

- [X] Completed

### Describe the feature

VSCode and many other apps support setting two different themes for light mode and dark mode, while respecting the underlying preference of the OS - this allows things like having dark mode at night and light mode during the day.

It'd be great to have support for this!

### If applicable, add mockups / screenshots to help present your vision of the feature

```json5
// Zed settings
//
// For information on how to configure Zed, see the Zed
// documentation: https://zed.dev/docs/configuring-zed
//
// To see all of Zed's default settings without changing your
// custom settings, run the `open default settings` command
// from the command palette or from `Zed` application menu.
{
    "theme": {
        "light": "One Light",
        "dark": "One Dark"
    },
    // "buffer_font_size": 15,
    // "language_overrides": {
    //     "JavaScript": {
    //         "format_on_save": {
    //             "external": {
    //                 "command": "npx prettier",
    //                 "arguments": [
    //                     "--stdin-filepath",
    //                     "{buffer_path}"
    //                 ]
    //             }
    //         }
    //     }
    // }
}
```

### LINKED ISSUE #5326: Auto switching light/dark theme according to system settings
Please, implement a setting to auto switch light/dark theme according to system settings.

Thank you!

### PR DESCRIPTION
**Pull request by Yesterday17**
Resolves #4970
Resolves #5326 

## Checklist

- [x] Added `themes` configuration
- [x] ~~Added an action to toggle between different theme modes~~ Removed as `auto` mode has been implemented
- [x] Added appearance observer to GPUI
- [x] Subscribe appearance change and change theme

### CODE REVIEW

**Comment by Yesterday17** [comment]
@cla-bot check

**Comment by TheDarkchip** [comment]
When the editor is launched in a mode that is not equal to the system appearance, it won't change themes. Only when the system appearance transition is triggered the theme is switched.

Maybe we can introduce a callback on the launch of the editor that checks for the current system appearance?

File: crates/gpui/src/window.rs
Comment (mikayla-maki) [inline crates/gpui/src/window.rs:393]: Oh wow, I hadn't realized we had already wired up the platform callback! :D

**Review by mikayla-maki — DISMISSED** [review]
This PR is awesome! I did a few tiny touch ups and adjusted how the theme is reloaded. As soon as CI passes, I'll merge this. Great work :)

**Comment by maxdeviant** [comment]
> This PR is awesome! I did a few tiny touch ups and adjusted how the theme is reloaded. As soon as CI passes, I'll merge this. Great work :)

@mikayla-maki Let's hold off on merging this one until I figure out how it works into the theme stuff I'm working on now.

Want to make sure we have a cohesive narrative around theming.

File: crates/theme/src/settings.rs
Comment (maxdeviant) [inline crates/theme/src/settings.rs:35]: Do we need to pass in `system_is_dark` if we can read it from the context?

It looks like at the call sites we always just pass in `cx.is_dark_mode()`.

File: crates/theme/src/settings.rs
Comment (maxdeviant) [inline crates/theme/src/settings.rs:115]: I think it would be nice if we could have these settings live under the same `theme` key, since they're mutually-exclusive with it.

This should be possible by introducing a new enum to use as the value for `theme`:

```
#[serde(untagged)]
enum ThemeNameOrDynamic {
    Theme(String),
    Dynamic(ThemeFeatures)
}
```

File: crates/theme/src/settings.rs
Comment (maxdeviant) [inline crates/theme/src/settings.rs:152]: Instead of having `is_dark_mode`, could we have an `appearance` method that returns an [`Appearance`](https://github.com/zed-industries/zed/blob/9cb5a84b8d37062bdd0f0afe7aeae64f714d1568/crates/theme/src/theme.rs#L34-L37)?

File: crates/theme/src/settings.rs
Comment (maxdeviant) [inline crates/theme/src/settings.rs:156]: Any reason to not just use `None` here instead of a wildcard? It looks like that should be the only missing case:

```suggestion
            Some(ThemeMode::Light) | None => false,
```

File: crates/theme/src/settings.rs
Comment (maxdeviant) [inline crates/theme/src/settings.rs:268]: I think storing an `Appearance` here would be clearer than storing a `bool`:

```suggestion
struct SystemAppearance(Appearance);
```

**Comment by TheDarkchip** [comment]
While the detection on editor startup now works, there is some trouble left with more than one system theme change.

Notably the themes.list() of the ThemeRegistry returns only one theme (One Dark).
As follows the .get(name) method also fails on a differing name, resulting in no theme change, even when the system theme changes.

**Comment by maxdeviant** [comment]
> I think I'm going to get this shipped @Yesterday17, thank you for getting the ball rolling :D

@mikayla-maki I was thinking it'd be nice to get the GPUI changes landed first, as those seem pretty independent.

And then I'd be down to pair on how we expose this in the settings, as I think there's some refinement we could do there.

**Comment by mikayla-maki** [comment]
@maxdeviant which GPUI changes are you referring to?

**Comment by maxdeviant** [comment]
@Yesterday17 Thank you for your work on this!

We ended up taking a slightly different approach with #7404, but we appreciate your jumping in and getting the ball rolling on this feature 🙂

**Pull request closed without merging**
